### ISSUE TITLE
Mapeo Desktop Installer doesn't run on 32bit machines 

### ISSUE THREAD

**Issue by arky** [digidem/mapeo-desktop#305]
Trying to run the Mapeo Desktop (5.x) installer on 32 bit Windows 10 Pro machines results in '64-bit Windows is required' message.  I have tested multiple versions of latest installers. 

![DSC_0149](https://user-images.githubusercontent.com/808895/76473900-7e635880-642c-11ea-8036-4c4601d8bff4.JPG)
![DSC_0150](https://user-images.githubusercontent.com/808895/76474005-d00be300-642c-11ea-8379-4865a136b372.JPG)

Most NGO/CSO organization have legacy machines hence the need for 32bit installers. I had managed to download and install Atom editor installer on this machine to check if electronJS applications are supported.

**Comment by arky** [comment]
@okdistribute  Testing windows artifact from [build 5843279](https://github.com/digidem/mapeo-desktop/actions/runs/58438279) results in following error.
![DSC_0001](https://user-images.githubusercontent.com/808895/77065690-58354e00-6a14-11ea-947e-22b6666b74fc.JPG)

**Comment by okdistribute** [comment]
I was able to get more information by using `asar: false` in the windows build:

Looks like the issue is with leveldown. We are one of the 1.9% here that need to create our own build.. https://github.com/Level/leveldown/issues/554
![Screenshot from 2020-03-19 20-54-40](https://user-images.githubusercontent.com/633012/77134373-e5cc6a00-6a23-11ea-82ad-10f528332a71.png)

**Comment by arky** [comment]
@okdistribute Yup, That reminds me that I had to rebuild leveldown package before building mapeo-desktop on 32bit machines.

**Comment by okdistribute** [comment]
@arky what did you type to rebuild?

**Comment by arky** [comment]
To build mapeo-desktop on 32bit machine you had rebuild leveldown without 64-bit flag. 

`rebuild-leveldb": "cd node_modules/leveldown && cross-env HOME=~/.electron-gyp node-gyp rebuild --target=2.0.7 --runtime=electron --dist-url=https://atom.io/download/atom-shell`

Please refer to this bug https://github.com/digidem/mapeo-desktop/issues/237

**Comment by vweevers** [comment]
Which version of leveldown are y'all on? The above command suggests on older version.

On latest leveldown (5.x) there should be no need to rebuild. If node is 32-bit, then on `npm install` leveldown will see that it has no bundled prebuild for windows 32-bit and fallback to compilation. The resulting build is compatible with both node and electron.

**Comment by okdistribute** [comment]
@vweevers Thanks for responding. I realize now that you're right, we updated to leveldown 5.x since #237 was filed. 

That's what I thought but still running into issues... I've been doing this on GitHub Actions and [experimenting on this PR](https://github.com/digidem/mapeo-desktop/pull/311)

**Comment by okdistribute** [comment]
@vweevers okay, doing this it seems to not have rebuilt leveldown, and says it can't find a native build for win32 
![Screenshot from 2020-03-20 11-29-23](https://user-images.githubusercontent.com/633012/77194929-10f39f80-6a9e-11ea-9b10-9c0ee1b714f1.png)

I noticed here in the [log for this build](
https://github.com/digidem/mapeo-desktop/runs/522624319?check_suite_focus=true#step:8:27) that electron-builder is rebuilding native modules for leveldown, but I wonder if that's failing silently?

**Comment by vweevers** [comment]
Could be, because those electron tools are geared towards the older `node-pre-gyp` and `prebuild-install`. I don't know if they support leveldown's setup.

I've got a Windows machine and can maybe help, given some instructions on how to get started on `mapeo-desktop`.

**Comment by okdistribute** [comment]
@vweevers thanks, that would be great. We're a small team and non-profit so any help is greatly appreciated. Mapeo Desktop should work out of the box with `npm  install` and `npm run dev` https://github.com/digidem/mapeo-desktop/#getting-started

**Comment by vweevers** [comment]
With:

```
$ node -v
v12.11.1
$ node -p process.arch
ia32
$ git branch
* master
```

I see that `npm install` correctly compiles `leveldown`.

There's also a postinstall script that (deep breath) runs `electron-builder install-app-deps` which in turn spawns `node_modules/app-builder-bin/win/ia32/app-builder.exe rebuild-node-modules` which is [written in Go](https://github.com/develar/app-builder/blob/master/pkg/node-modules/rebuild.go) and effectively spawns `npm rebuild leveldown@5.5.1 ...` which calls the `install` script of `leveldown`. Which doesn't do anything, because it previously compiled and doesn't check whether that build is compatible with the target runtime. I'll get back to that.

TLDR, `electron-builder install-app-deps` doesn't do anything (at least for `leveldown`) but at this point in time we still have a `leveldown` build that happens to be compatible with Electron.

The `npm run build` and `npm run dev` commands also work, and I see a LevelDB db getting created on disk, so `leveldown` is correctly loaded.

Next up is packaging; I'll try that in your `32bit-builds` branch.

**Comment by vweevers** [comment]
@okdistribute After running `npm run dist` (in the `32bit-builds` branch) how can I launch Mapeo? Can I simply open `dist/win-ia32-unpacked/Mapeo.exe` or is that not a realistic test and should I instead run the installer exe?

Side note: the `32bit-builds` branch has this (overly helpful) error:

```
$ npx electron-builder --ia32
  • electron-builder  version=21.2.0 os=10.0.18362
  • loaded configuration  file=package.json ("build" field)
  ⨯ Package "electron-rebuild" is only allowed in "devDependencies". Please remove it 
      from the "dependencies" section in your package.json.
```

**Comment by vweevers** [comment]
The 32-bit leveldown build correctly gets included in `win-ia32-unpacked/resources/app/node_modules/..` and `Mapeo.exe` works - but it created registry keys so I ran the installer exe anyway to cleanup those keys.

And... the installed Mapeo also works!

**Comment by vweevers** [comment]
Given the amount of tools involved, it might take some time and thought to properly fix rebuilding. So I'm leaning towards just adding a 32-bit prebuild to `leveldown` (and making note of the issues elsewhere).

**Comment by okdistribute** [comment]
Thanks so much for investigating. This was my worry -- that we'd have to create separate builds for each 32 and 64 bit systems. This isn't really an issue since most users us 64 bit, but sometimes we don't know what kind of system a user will have until we go out to help them install the app. 

Adding 32-bit builds to leveldown will really help the electron ecosystem!

**Comment by okdistribute** [comment]
This will be fixed in the next release, but you can try it now in this build https://github.com/digidem/mapeo-desktop/actions/runs/67706742

**Comment by arky** [comment]
@okdistribute  Tested on windows 32 bit machine. Works like a charm.

**Comment by vweevers** [comment]
> Given the amount of tools involved, it might take some time and thought to properly fix rebuilding.

Update (FYI): `node-gyp-build` now respects the `electron-builder` [`buildDependenciesFromSource`](https://www.electron.build/configuration/configuration#Configuration-buildDependenciesFromSource) option (https://github.com/prebuild/node-gyp-build/pull/30).

**Comment by okdistribute** [comment]
nice thanks @vweevers !
